spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/AbstractLocalDeployerSupport.java

```diff
@@ -384,23 +393,42 @@ else if(request.getDefinition().getProperties().containsKey(LocalAppDeployer.SER
 		return port;
 	}
 
-	private int getRandomPort(int port) {
-		int randomInt = ThreadLocalRandom.current().nextInt(50000, 61000);
-		Set<Integer> availPorts = SocketUtils.findAvailableTcpPorts(10, randomInt, randomInt+10);
-		logger.debug("Available Ports: " + availPorts.toString());
-		Random rand = new Random(System.currentTimeMillis());
-		int randomIndex = rand.nextInt(availPorts.size());
-		int i = 0;
-		for(Integer freePort : availPorts)
-		{
-			if (i == randomIndex) {
-				port = freePort;
+	public synchronized int getRandomPort() {
+		Set<Integer> availPorts = new HashSet<>();
+		// SocketUtils.findAvailableTcpPorts retries 6 times, add additional retry on top.
+		for (int retryCount = 0; retryCount < 5; retryCount++) {
+			int randomInt = ThreadLocalRandom.current().nextInt(20000, 61000);
+			try {
+				availPorts = SocketUtils.findAvailableTcpPorts(5, randomInt, randomInt + 5);
+				try {
+					// Give some time for the system to release up ports that were scanned.
+					Thread.sleep(100);
```

Comment:
I don't think we need to sleep here. If this is because I commented something it was just a theoretical question.

Reply:
I had a more instrumented test harness counting all the ways stuff could fail, I saw the failure rate go down when the sleep was added.

Reply:
Well that sucks. It means that this concept is even more unreliable what I thought.